src/carousel/index.ts

```diff
@@ -0,0 +1,206 @@
+    this.animationEndCallback = null;
+  }
+
+  resetCurrentSlideNum = (from: number, to: number, length: number) => {
+    const currentSlide = from < to ? to % length : (to % length + length) % length;
+    this.set({ currentSlide });
+  }
+
+  disableTransition = () => {
+    this.refs.track.removeEventListener('transitionend', this.disableTransition, false);
+    this.set({ transitioning: false });
+  }
+
+  trackStyle = () => {
+    const { speed } = this.props;
+    if (this.state.transitioning && speed) {
+      const transition = `${speed}ms ease`;
+      return {
+        ...this.getStaticTrackStyle(),
+        '-webkit-transition': transition,
+        transition,
+        '-ms-transition': transition
+      };
+    } else {
+      return this.getStaticTrackStyle();
+    }
+  }
+
+  getStaticTrackStyle = () => {
+    const slideWidth = this.getSlideWidth();
+    const slideCount = this.props.items.length;
+    const slidestoshow = this.props.slidestoshow;
+    const trackWidth = (slideCount + 4 * slidestoshow - 2) * slideWidth;
+    const pos = Carousel.calculatePosition(this.state.currentSlide, slideWidth, slidestoshow);
+    const transform = `translateX(-${pos}px)`;
+
+    return {
+      width: `${trackWidth}px`,
+      transform,
+      '-webkit-transform': transform,
+      '-ms-transform': transform
+    };
+  }
+
+  getSlideWidth = () => this.refs.wrapper.offsetWidth / this.props.slidestoshow;
+
+  onTouchStart = (e: TouchEvent & Carousel.Event) => {
+    this.state.touchObject = {
+      startX: e.touches[0].pageX,
+      startY: e.touches[0].pageY
+    };
+
+    e.target.addEventListener('touchend', this.onTouchEnd);
+  }
+
+  onTouchEnd = (e: TouchEvent & Carousel.Event) => {
+    const curX = e.changedTouches[0].pageX;
+    const curY = e.changedTouches[0].pageY;
+    const swipeNext = Carousel.isSwipeToNext({ ...this.state.touchObject }, curX, curY);
+
+    // swipe distance needs to be more than 20
+    if (Math.abs(curX - this.state.touchObject.startX) > MIN_SWIPE_DISTANCE && typeof swipeNext === 'boolean') {
+      swipeNext ? this.moveNext() : this.movePrevious();
+    }
+
+    e.target.removeEventListener('touchstart', this.onTouchStart);
+    e.target.removeEventListener('touchend', this.onTouchEnd);
+  }
+
+  static calculatePosition = (currentSlide: number, moveDistance: number, slidestoshow: number): number =>
+    (currentSlide + slidestoshow * 2 - 1) * moveDistance
+
+  static isSwipeToNext = (touchObj: Carousel.TouchObject, curX: number, curY: number): boolean => {
+    const xDist = touchObj.startX - curX;
+    const yDist = touchObj.startY - curY;
+
+    const r = Math.atan2(yDist, xDist);
+
+    let swipeAngle = Math.round(r * 180 / Math.PI);
+
+    if (swipeAngle < 0) {
+      swipeAngle = 360 - Math.abs(swipeAngle);
+    }
+
+    // swipeAngle between 45 and 135, between 225 and 315 is ignored;
+    if ((swipeAngle <= 45 && swipeAngle >= 0) || (swipeAngle <= 360 && swipeAngle >= 315)) {
```

Comment:
can we save `45`, `135` and `225` as `const`s at the top of this file for clarity around what they represent